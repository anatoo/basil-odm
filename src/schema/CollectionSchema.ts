import { CollectionDef, Index } from '../types';
import { createFieldsSchema, FieldsSchema } from './FieldsSchema';
import { ObjectSchemaSource } from './types';

interface Props {
  collectionName: string;
  fields?: ObjectSchemaSource;
  indexes?: Index[];
  entityName?: string;
}

export class CollectionSchema implements CollectionDef {
  _collectionName: string;
  _fields: FieldsSchema;
  _indexes: Index[];
  _entityName: string | null = null;

  constructor(props: Props) {
    const { collectionName, fields = {}, indexes = [], entityName = null } = props;

    this._collectionName = collectionName;
    this._fields = createFieldsSchema(fields);
    this._indexes = indexes;
    this._entityName = entityName;
  }

  get collectionName(): string {
    return this._collectionName;
  }

  get entityName(): string | null {
    return this._entityName;
  }

  get schema(): FieldsSchema {
    return this._fields;
  }

  get indexes(): Index[] {
    return this._indexes;
  }
}
