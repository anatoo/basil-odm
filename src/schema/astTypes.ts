/**
 * スキーマを表現するASTを定義する
 *
 * 目的:
 *  - インスペクション可能なデータ構造を提供する
 *  - スキーマを変換したり操作したりできるようにする
 *  - 様々なアウトプットに対応できるようにする(例: json-schema, TypeScriptの型、bson-schema、その他コードなど)
 *  - コレクションのドキュメント同士のリレーションを定義できるようにする
 **/

export type SchemaRoot = ObjectNode;

export type SchemaNode = ObjectNode | ArrayNode | Union | Value | Literal | Enum | Reference;

export type ObjectNode = {
  kind: 'object';
  props: {
    [key: string]: Field;
  };
  allowAdditionalProps: boolean;
};

export type Field = {
  kind: 'field';
  isOptional: boolean;
  node: SchemaNode;
};

// 他のドキュメントを参照する
export type Reference = {
  kind: 'reference';
  node: StringValue | ObjectIdValue;
  collectionName?: string; // 省略された場合には、自身のコレクションを参照する
  path: string; // field名をdotで繋げたもの。 例: 'workspace.id'
};

export type ArrayNode = {
  kind: 'array';
  item: SchemaNode;
};

export type Union = {
  kind: 'union';
  items: SchemaNode[];
};

export type LiteralValue = string | boolean | null | number;

export type Literal = {
  kind: 'literal';
  value: LiteralValue;
};

export type Enum = {
  kind: 'enum';
  values: Record<string, LiteralValue>;
  name?: string;
};

export type Value = StringValue | NumberValue | ObjectIdValue | NullValue | BooleanValue | DateValue | TimestampValue | BinaryValue;

export type StringValue = {
  kind: 'string';
  maxLength?: number;
};

export type NumberValue = {
  kind: 'number';
};

export type ObjectIdValue = {
  kind: 'objectId';
};

export type NullValue = {
  kind: 'null';
};

export type BooleanValue = {
  kind: 'boolean';
};

export type DateValue = {
  kind: 'date';
};

export type TimestampValue = {
  kind: 'timestamp';
};

export type BinaryValue = {
  kind: 'binary';
};

// TODO decimal, int, doubleあたりのデータ型を追加する
// TODO numberを無くす
// TODO 数値型の場合には、制限を設けるパラメータを付ける(範囲など)
