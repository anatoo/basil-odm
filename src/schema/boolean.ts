import { SchemaFragment } from './types';
import { schemaFragmentFrag } from './symbols';

function originalBoolean(): SchemaFragment {
  return {
    [schemaFragmentFrag]: true,

    buildASTNode() {
      return {
        kind: 'boolean',
      };
    },
  };
}

export const boolean: typeof originalBoolean & SchemaFragment = Object.assign(originalBoolean, originalBoolean());
