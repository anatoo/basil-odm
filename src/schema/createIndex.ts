import { IndexFields, Index } from '../types';
import mongodb from 'mongodb';

export function index(fields: IndexFields, options: mongodb.CreateIndexesOptions = {}): Index {
  return {
    fields,
    options,
  };
}
