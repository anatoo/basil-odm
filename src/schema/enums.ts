import { schemaFragmentFrag } from './symbols';
import { SchemaFragment } from './types';
import { inspect } from 'util';

export function enums(props: { values: Record<string, Value> | Record<number, Value> }): SchemaFragment {
  const { values } = props;

  return new Enum({
    values: Array.isArray(values)
      ? values.reduce((record, item, index) => {
          record['' + index] = item;
          return record;
        }, {})
      : values,
  });
}

type Value = string | boolean | number | null;

interface Props {
  values: Record<string, Value>;
}

class Enum implements SchemaFragment {
  values: Record<string, Value>;

  [schemaFragmentFrag]: true = true as const;

  constructor(props: Props) {
    const { values } = props;

    Object.values(values).forEach((value) => {
      if (typeof value !== 'string' && typeof value !== 'boolean' && typeof value !== 'number' && value !== null) {
        throw Error(`Enum values must be a string, boolean, number, or null. Got "${inspect(value)}".`);
      }
    });

    this.values = values;
  }

  buildASTNode() {
    return {
      kind: 'enum' as const,
      values: this.values,
    };
  }
}
