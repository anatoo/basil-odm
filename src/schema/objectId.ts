import { SchemaFragment } from './types';
import { schemaFragmentFrag } from './symbols';

function originalObjectId(): SchemaFragment {
  return {
    [schemaFragmentFrag]: true,

    buildASTNode() {
      return {
        kind: 'objectId',
      };
    },
  };
}

export const objectId: typeof originalObjectId & SchemaFragment = Object.assign(originalObjectId, originalObjectId());
