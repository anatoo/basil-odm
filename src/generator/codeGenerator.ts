import { FieldsSchemaRoot, Enum } from '../schema/astTypes';
import { CollectionDef, DefinedSchema } from '../types';
import { writeFileSync } from 'fs';
import { prettier } from '../utils';
import { aggregateEnums, generateDefaultValue, generateType, getDefaultEntityName } from './utils';
import { inspect } from 'util';

export const header = `/**
 * This file was generated by Basil ODM
 * Do not make changes to this file directly
 */
/* eslint-disable */
import * as mongodb from 'mongodb';
import * as basil from 'basil-odm';

`;

export function generateEnumsCode(collections: CollectionDef[]): string {
  let code = '';

  let enums = collections.flatMap((collection) => {
    const root = collection.schema.getSchemaAST();
    return aggregateEnums(root);
  });

  // remove duplicated
  const enumsMap = new Map<string, Enum>();
  for (const node of enums) {
    if (typeof node.name === 'string') {
      if (enumsMap.has(node.name)) {
        const storedNode = enumsMap.get(node.name) as Enum;
        if (storedNode !== node && JSON.stringify(storedNode) !== JSON.stringify(node)) {
          throw Error(`These enums node has same name but different object: ${inspect(node)} ${inspect(storedNode)}`);
        }
      } else {
        enumsMap.set(node.name, node);
      }
    }
  }

  enums = [...enumsMap.values(), ...enums.filter((node) => typeof node.name !== 'string')];

  for (const node of enums) {
    if (node.name) {
      code += `export const ${node.name} = {`;
      for (const [key, value] of Object.entries(node.values)) {
        code += `${JSON.stringify(key)}: ${JSON.stringify(value)},`;
      }
      code += `} as const;\n`;

      code += `export type ${node.name} = typeof ${node.name}[keyof typeof ${node.name}];\n\n`;
    }
  }

  return code;
}

export function generateTypeScriptFile(collections: CollectionDef[]): string {
  return header + generateCollectionDefs(collections) + generateDocumentTypes(collections) + generateEnumsCode(collections);
}

export function generateCollectionDefs(collections: CollectionDef[]): string {
  let code = '';

  code += `const $defs: Record<string, basil.EntityMeta<Record<string, unknown>>> = {`;
  for (const collection of collections) {
    code += `
${JSON.stringify(collection.collectionName)}: {
  collectionName: ${JSON.stringify(collection.collectionName)},
  schema: new basil.FieldsSchema(${JSON.stringify(collection.schema.getSchemaAST())}),
  indexes: ${JSON.stringify(collection.indexes)},
},`;
  }
  code += `};\n\n`;

  return code;
}

export function generateDocumentTypes(collections: CollectionDef[]): string {
  let code = ``;

  for (const collection of collections) {
    const ast = collection.schema.getSchemaAST();
    const name = collection.entityName ?? getDefaultEntityName(collection.collectionName);

    code += `export class ${name} extends basil.Base {
constructor(params?: Partial<${name}>) {
  super();
  Object.assign(this, params);
}
static getCollection() {
  return {
    collectionName: ${JSON.stringify(collection.collectionName)},
    schema: $defs[${JSON.stringify(collection.collectionName)}].schema,
    indexes: $defs[${JSON.stringify(collection.collectionName)}].indexes,
  };
}`;

    for (const [name, field] of Object.entries(ast.props)) {
      code += `${name}${field.isOptional ? '?' : ''}: ${generateType(field.node)}`;
      if (!field.isOptional) {
        code += '= ' + generateDefaultValue(field.node);
      }
      code += ';\n';
    }

    code += '}\n\n';
  }

  return code;
}

export function generateTypeFromSchema(root: FieldsSchemaRoot) {
  return generateType(root);
}

interface GenerateCodeParams {
  outputFile: string;
  schema: DefinedSchema;
}

export function generateCode(params: GenerateCodeParams) {
  const { schema, outputFile } = params;
  const collections = Object.values(schema);
  writeFileSync(outputFile, prettier(generateTypeScriptFile(collections)));
}
