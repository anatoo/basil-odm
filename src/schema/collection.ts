import { CollectionDef, Index, CollectionOptions } from '../types';
import { createFieldsSchema } from './FieldsSchema';
import { ObjectSchemaSource } from './types';
import { Edge } from './edgeTypes';
import { getDefaultEntityName } from '../generator/utils';
interface Props {
  collectionName: string;
  fields?: ObjectSchemaSource;
  indexes?: Index[];
  entityName?: string;
  options?: CollectionOptions;
  edges?: Record<string, Edge>;
}

export function collection(props: Props): CollectionDef {
  return {
    collectionName: props.collectionName,
    entityName: props.entityName || getDefaultEntityName(props.collectionName),
    fields: createFieldsSchema(props.fields || {}),
    indexes: props.indexes || [],
    options: props.options,
    edges: props.edges,
  };
}
