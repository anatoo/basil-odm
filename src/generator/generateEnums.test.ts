import { enums } from '../schema/enums';
import { generateEnumsCode } from './codeGenerator';
import { CollectionSchema } from '../schema/CollectionSchema';
import { objectId } from '../schema/objectId';

describe('generateEnums()', () => {
  test('normal', () => {
    const collection = new CollectionSchema({
      collectionName: 'users',
      entityName: 'User',
      fields: {
        _id: objectId,
        type: enums({
          name: 'UserType',
          values: ['member', 'admin'],
        }),
      },
    });

    expect(generateEnumsCode([collection])).toMatchSnapshot();
  });

  test('without name', () => {
    const collection = new CollectionSchema({
      collectionName: 'users',
      entityName: 'User',
      fields: {
        _id: objectId,
        type: enums({
          values: ['member', 'admin'],
        }),
      },
    });

    expect(generateEnumsCode([collection])).toMatchSnapshot();
  });
});
