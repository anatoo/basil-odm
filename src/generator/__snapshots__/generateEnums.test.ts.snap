// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateEnums() normal 1`] = `"export const UserType = {  "0": "member",  "1": "admin",} as const;"`;

exports[`generateEnums() without name 1`] = `""`;